File: CS-Examples/10_Conversion/ConvertObjectToImage.cs
using System;
using System.Windows.Forms;
using System.Text;
using Spire.Doc;
using Spire.Doc.Documents;
using System.Drawing;
using Spire.Doc.Formatting;
using System.IO;
using Spire.Doc.Fields;
using System.Drawing.Imaging;

namespace ConvertObjectToImage
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            //Create a document
			Document document = new Document();

			//Load the document from a file
			document.LoadFromFile(@"..\..\..\..\..\..\Data\ConvertObjectToImage.docx");

			//Get the first section
			Section section = document.Sections[0];

			//Get body of section
			Body body = section.Body;

			//Get the first paragraph
			Paragraph paragraph = body.Paragraphs[0];

			//Convert the paragraph to an image
			Image image = ConvertParagraphToImage(paragraph);

			//Save the image
			image.Save("ConvertParagraphToImage.png", ImageFormat.Png);

			//Get the first table
			Table table = body.Tables[0] as Table;

			//Convert the table to an image
			image = ConvertTableToImage(table);

			//Save the image
			image.Save("ConvertTableToImage.jpg", ImageFormat.Jpeg);

			//Get the first row of the first table
			TableRow row = table.Rows[0];
			image = ConvertTableRowToImage(row);

			//Save the image
			image.Save("ConvertTableRowToImage.bmp", ImageFormat.Bmp);

			//Get the first cell of the first row
			TableCell cell = row.Cells[0];
			image = ConvertTableCellToImage(cell);

			//Save the image
			image.Save("ConvertTableCellToImage.png", ImageFormat.Png);

			int i = 0;

			//Iterate over paragraphs in the section
			foreach (Paragraph p in section.Paragraphs)
			{
				// Iterate over child objects in the paragraph
				foreach (DocumentObject obj in p.ChildObjects)
				{
					//Check if the object is a shape
					if (obj.DocumentObjectType == DocumentObjectType.Shape)
					{
						//Convert the shape to an image
						image = ConvertShapeToImage(obj as ShapeObject);

						//Save the image
						image.Save(String.Format("ConvertShapeToImage-{0}.png", i), ImageFormat.Png);
						i++;
					}
				}
			}

			//Dispose the document
			document.Dispose();
           
        }
        
		private Image ConvertParagraphToImage(Paragraph obj)
		{
			//Create a new document
			Document doc = new Document();

			//Add a new section
			Section section = doc.AddSection();

			//Add a deep clone of the paragraph to the section
			section.Body.ChildObjects.Add(obj.Clone());

			//Save the image
			Image image = doc.SaveToImages(0, ImageType.Bitmap);

			//Close the document
			doc.Close();
			return CutImageWhitePart(image as Bitmap, 1);
		}
		private Image ConvertTableToImage(Table obj)
		{
			//Create a new document
			Document doc = new Document();

			//Add a section to the document
			Section section = doc.AddSection();

			//Add a deep clone of the table to the section
			section.Body.ChildObjects.Add(obj.Clone());

			//Save the image
			Image image = doc.SaveToImages(0, ImageType.Bitmap);

			//Close the document
			doc.Close();
			return CutImageWhitePart(image as Bitmap, 1);
		}
		
		private Image ConvertTableRowToImage(TableRow obj)
		{
			//Create a new document
			Document doc = new Document();

			//Add a section to the document
			Section section = doc.AddSection();

			//Add a table to the section
			Table table = section.AddTable();

			//Add a deep clone of the row to the table
			table.Rows.Add(obj.Clone());

			//Save the image
			Image image = doc.SaveToImages(0, ImageType.Bitmap);
			doc.Close();
			return CutImageWhitePart(image as Bitmap, 1);
		}

		private Image ConvertTableCellToImage(TableCell obj)
		{
			// Create a new document
			Document doc = new Document();

			//Add a section to the document
			Section section = doc.AddSection();

			//Add a table to the section
			Table table = section.AddTable();

			//Add a new row to the table and add a deep clone of the cell to it
			table.AddRow().Cells.Add(obj.Clone());

			//Save the image
			Image image = doc.SaveToImages(0, ImageType.Bitmap);
			doc.Close();
			return CutImageWhitePart(image as Bitmap, 1);
		}
		
		private Image ConvertShapeToImage(ShapeObject obj)
		{
			//Create a new document
			Document doc = new Document();

			//Add a section to the document
			Section section = doc.AddSection();

			// Add a paragraph to the section and add a deep clone of the shape object to it
			section.AddParagraph().ChildObjects.Add(obj.Clone());

			//Create a MemoryStream
			MemoryStream ms = new MemoryStream();

			//Save the document to stream
			doc.SaveToStream(ms, FileFormat.Docx);

			//Load a document from stream
			doc.LoadFromStream(ms, FileFormat.Docx);

			//Save to image
			Image image = doc.SaveToImages(0, ImageType.Bitmap);

			//Close the document and stream
			ms.Close();
			doc.Close();
			return CutImageWhitePart(image as Bitmap, 1);
		}
		public Image CutImageWhitePart(Bitmap bmp, int WhiteBarRate)
	{
		int top = 0, left = 0;
		int right = bmp.Width, bottom = bmp.Height;
		Color white = Color.White;

		for (int i = 0; i < bmp.Height; i++)
		{
			bool find = false;
			for (int j = 0; j < bmp.Width; j++)
			{
				Color c = bmp.GetPixel(j, i);
				if (IsWhite(c))
				{
					top = i;
					find = true;
					break;
				}
			}
			if (find) break;
		}

		for (int i = 0; i < bmp.Width; i++)
		{
			bool find = false;
			for (int j = top; j < bmp.Height; j++)
			{
				Color c = bmp.GetPixel(i, j);
				if (IsWhite(c))
				{
					left = i;
					find = true;
					break;
				}
			}
			if (find) break; ;
		}

		for (int i = bmp.Height - 1; i >= 0; i--)
		{
			bool find = false;
			for (int j = left; j < bmp.Width; j++)
			{
				Color c = bmp.GetPixel(j, i);
				if (IsWhite(c))
				{
					bottom = i;
					find = true;
					break;
				}
			}
			if (find) break;
		}

		for (int i = bmp.Width - 1; i >= 0; i--)
		{
			bool find = false;
			for (int j = 0; j <= bottom; j++)
			{
				Color c = bmp.GetPixel(i, j);
				if (IsWhite(c))
				{
					right = i;
					find = true;
					break;
				}
			}
			if (find) break;
		}
		int iWidth = right - left;
		int iHeight = bottom - left;
		int blockWidth = Convert.ToInt32(iWidth * WhiteBarRate / 100);
		bmp = Cut(bmp, left - blockWidth, top - blockWidth, right - left + 2 * blockWidth, bottom - top + 2 * blockWidth);

		return bmp;

	}
		public Bitmap Cut(Bitmap b, int StartX, int StartY, int iWidth, int iHeight)
		{
			if (b == null)
			{
				return null;
			}
			int w = b.Width;
			int h = b.Height;
			if (StartX >= w || StartY >= h)
			{
				return null;
			}
			if (StartX + iWidth > w)
			{
				iWidth = w - StartX;
			}
			if (StartY + iHeight > h)
			{
				iHeight = h - StartY;
			}
			try
			{
				Bitmap bmpOut = new Bitmap(iWidth, iHeight, PixelFormat.Format24bppRgb);
				Graphics g = Graphics.FromImage(bmpOut);
				g.DrawImage(b, new Rectangle(0, 0, iWidth, iHeight), new Rectangle(StartX, StartY, iWidth, iHeight), GraphicsUnit.Pixel);
				g.Dispose();
				return bmpOut;
			}
			catch
			{
				return null;
			}
		}


		public bool IsWhite(Color c)
		{
			if (c.R < 245 || c.G < 245 || c.B < 245)
				return true;
			else return false;
		}
	}
}

File: VB-Examples/10_Conversion/ConvertObjectToImage.vb
Imports System.Text
Imports Spire.Doc
Imports Spire.Doc.Documents
Imports Spire.Doc.Formatting
Imports System.IO
Imports Spire.Doc.Fields
Imports System.Drawing.Imaging

Namespace ConvertObjectToImage
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
			
			' Create a new document object
			Dim document As New Document()

			' Load the source document from file
			document.LoadFromFile("..\..\..\..\..\..\Data\ConvertObjectToImage.docx")

			' Get the first section of the document
			Dim section As Section = document.Sections(0)

			' Get the body of the section
			Dim body As Body = section.Body

			' Convert the first paragraph in the body to an image
			Dim paragraph As Paragraph = body.Paragraphs(0)
			Dim image As Image = ConvertParagraphToImage(paragraph)
			image.Save("ConvertParagraphToImage.png", ImageFormat.Png)

			' Convert the first table in the body to an image
			Dim table As Table = TryCast(body.Tables(0), Table)
			image = ConvertTableToImage(table)
			image.Save("ConvertTableToImage.jpg", ImageFormat.Jpeg)

			' Convert the first row in the table to an image
			Dim row As TableRow = table.Rows(0)
			image = ConvertTableRowToImage(row)
			image.Save("ConvertTableRowToImage.bmp", ImageFormat.Bmp)

			' Convert the first cell in the row to an image
			Dim cell As TableCell = row.Cells(0)
			image = ConvertTableCellToImage(cell)
			image.Save("ConvertTableCellToImage.png", ImageFormat.Png)

			' Iterate through paragraphs in the section and convert shape objects to images
			Dim i As Integer = 0
			For Each p As Paragraph In section.Paragraphs
				For Each obj As DocumentObject In p.ChildObjects
					If obj.DocumentObjectType = DocumentObjectType.Shape Then
						image = ConvertShapeToImage(TryCast(obj, ShapeObject))
						image.Save(String.Format("ConvertShapeToImage-{0}.png", i), ImageFormat.Png)
						i += 1
					End If
				Next obj
			Next p

			' Dispose the document
			document.Dispose()
		End Sub
		Private Function ConvertParagraphToImage(ByVal obj As Paragraph) As Image
			' Create a new document and section
			Dim doc As New Document()
			Dim section As Section = doc.AddSection()

			' Add the cloned paragraph to the section's body
			section.Body.ChildObjects.Add(obj.Clone())

			' Save the document to an image and close it
			Dim image As Image = doc.SaveToImages(0, ImageType.Bitmap)
			doc.Close()

			' Cut the white parts from the image
			Return CutImageWhitePart(TryCast(image, Bitmap), 1)
		End Function
		
		Private Function ConvertTableToImage(ByVal obj As Table) As Image
			' Create a new document
			Dim doc As New Document()
			' Add a section to the document
			Dim section As Section = doc.AddSection()

			' Add a clone of the table object to the section's body
			section.Body.ChildObjects.Add(obj.Clone())

			' Save the document as an image and retrieve the image object
			Dim image As Image = doc.SaveToImages(0, ImageType.Bitmap)

			' Close the document
			doc.Close()

			' Return the image after cutting the white parts
			Return CutImageWhitePart(TryCast(image, Bitmap), 1)
		End Function

		Private Function ConvertTableRowToImage(ByVal obj As TableRow) As Image
			' Create a new document
			Dim doc As New Document()
			' Add a section to the document
			Dim section As Section = doc.AddSection()
			' Add a table to the section
			Dim table As Table = section.AddTable()

			' Add a clone of the table row object to the table
			table.Rows.Add(obj.Clone())

			' Save the document as an image and retrieve the image object
			Dim image As Image = doc.SaveToImages(0, ImageType.Bitmap)

			' Close the document
			doc.Close()

			' Return the image after cutting the white parts
			Return CutImageWhitePart(TryCast(image, Bitmap), 1)
		End Function

		Private Function ConvertTableCellToImage(ByVal obj As TableCell) As Image
			' Create a new document
			Dim doc As New Document()
			' Add a section to the document
			Dim section As Section = doc.AddSection()
			' Add a table to the section
			Dim table As Table = section.AddTable()

			' Add a clone of the table cell object to a new row in the table
			table.AddRow().Cells.Add(obj.Clone())

			' Save the document as an image and retrieve the image object
			Dim image As Image = doc.SaveToImages(0, ImageType.Bitmap)

			' Close the document
			doc.Close()

			' Return the image after cutting the white parts
			Return CutImageWhitePart(TryCast(image, Bitmap), 1)
		End Function

		Private Function ConvertShapeToImage(ByVal obj As ShapeObject) As Image
			' Create a new document
			Dim doc As New Document()
			' Add a section to the document
			Dim section As Section = doc.AddSection()
			' Add a paragraph to the section's body
			section.AddParagraph().ChildObjects.Add(obj.Clone())

			' Save the document to a memory stream
			Dim ms As New MemoryStream()
			doc.SaveToStream(ms, FileFormat.Docx)

			' Load the document from the memory stream
			doc.LoadFromStream(ms, FileFormat.Docx)

			' Save the document as an image and retrieve the image object
			Dim image As Image = doc.SaveToImages(0, ImageType.Bitmap)

			' Close the memory stream
			ms.Close()
			' Close the document
			doc.Close()

			' Return the image after cutting the white parts
			Return CutImageWhitePart(TryCast(image, Bitmap), 1)
		End Function

		Public Function CutImageWhitePart(ByVal bmp As Bitmap, ByVal WhiteBarRate As Integer) As Image
			' Initialize variables for the boundaries of the white part
			Dim top As Integer = 0, left As Integer = 0
			Dim right As Integer = bmp.Width, bottom As Integer = bmp.Height
			Dim white As Color = Color.White

			' Find the top boundary of the white part
			For i As Integer = 0 To bmp.Height - 1
				Dim find As Boolean = False
				For j As Integer = 0 To bmp.Width - 1
					Dim c As Color = bmp.GetPixel(j, i)
					If IsWhite(c) Then
						top = i
						find = True
						Exit For
					End If
				Next j
				If find Then
					Exit For
				End If
			Next i

			' Find the left boundary of the white part
			For i As Integer = 0 To bmp.Width - 1
				Dim find As Boolean = False
				For j As Integer = top To bmp.Height - 1
					Dim c As Color = bmp.GetPixel(i, j)
					If IsWhite(c) Then
						left = i
						find = True
						Exit For
					End If
				Next j
				If find Then
					Exit For
				End If
			Next i

			' Find the bottom boundary of the white part
			For i As Integer = bmp.Height - 1 To 0 Step -1
				Dim find As Boolean = False
				For j As Integer = left To bmp.Width - 1
					Dim c As Color = bmp.GetPixel(j, i)
					If IsWhite(c) Then
						bottom = i
						find = True
						Exit For
					End If
				Next j
				If find Then
					Exit For
				End If
			Next i

			' Find the right boundary of the white part
			For i As Integer = bmp.Width - 1 To 0 Step -1
				Dim find As Boolean = False
				For j As Integer = 0 To bottom
					Dim c As Color = bmp.GetPixel(i, j)
					If IsWhite(c) Then
						right = i
						find = True
						Exit For
					End If
				Next j
				If find Then
					Exit For
				End If
			Next i

			' Calculate dimensions for cropping the image
			Dim iWidth As Integer = right - left
			Dim iHeight As Integer = bottom - top
			Dim blockWidth As Integer = Convert.ToInt32(iWidth * WhiteBarRate \ 100)

			' Crop the image based on the calculated boundaries and block width
			bmp = Cut(bmp, left - blockWidth, top - blockWidth, right - left + 2 * blockWidth, bottom - top + 2 * blockWidth)

			Return bmp
		End Function

		' Function to crop an image
		Public Function Cut(ByVal b As Bitmap, ByVal StartX As Integer, ByVal StartY As Integer, ByVal iWidth As Integer, ByVal iHeight As Integer) As Bitmap
			' Check if input bitmap is valid
			If b Is Nothing Then
				Return Nothing
			End If
			Dim w As Integer = b.Width
			Dim h As Integer = b.Height

			' Check if starting coordinates are within bounds
			If StartX >= w OrElse StartY >= h Then
				Return Nothing
			End If

			' Adjust dimensions if exceeding the image size
			If StartX + iWidth > w Then
				iWidth = w - StartX
			End If
			If StartY + iHeight > h Then
				iHeight = h - StartY
			End If

			Try
				' Create a new bitmap with the specified dimensions
				Dim bmpOut As New Bitmap(iWidth, iHeight, PixelFormat.Format24bppRgb)
				Dim g As Graphics = Graphics.FromImage(bmpOut)

				' Draw the cropped portion onto the new bitmap
				g.DrawImage(b, New Rectangle(0, 0, iWidth, iHeight), New Rectangle(StartX, StartY, iWidth, iHeight), GraphicsUnit.Pixel)
				g.Dispose()
				Return bmpOut
			Catch
				Return Nothing
			End Try
		End Function

		' Function to check if a color is considered white
		Public Function IsWhite(ByVal c As Color) As Boolean
			' Check if any RGB component of the color is less than 245 (on a scale of 0-255)
			If c.R < 245 OrElse c.G < 245 OrElse c.B < 245 Then
						Return True
					Else
						Return False
					End If
				End Function
	End Class
End Namespace
